csharp/src/Ice/IncomingRequestFrame.cs

```diff
@@ -3,39 +3,66 @@
 //
 
 using System;
+using System.Collections.Generic;
 
 namespace Ice
 {
     /// <summary>Represents a request protocol frame received by the application.</summary>
-    // TODO: IncomingRequestFrame should derive from InputStream
     public sealed class IncomingRequestFrame
     {
-        public InputStream InputStream { get; }
+        public IReadOnlyDictionary<string, string> Context { get; }
+        public Encoding Encoding { get; }
+        public string Facet { get; }
+        public Identity Identity { get; }
+        public bool IsIdempotent { get; }
+        public string Operation { get; }
 
-        // TODO: combine Current and RequestFrameHeader in some way?
-        public Current Current { get; }
+        public int Size => Data.Count;
 
         /// <summary>The payload of this request frame. The bytes inside the payload should not be written to;
         /// they are writable because of the <see cref="System.Net.Sockets.Socket"/> methods for sending.</summary>
         // TODO: describe how long this payload remains valid once we add memory pooling.
         public ArraySegment<byte> Payload { get; }
 
-        /// <summary>Take the payload from this response frame. After calling this method, the payload can no longer
-        /// be read.</summary>
-        public ArraySegment<byte> TakePayload()
+        internal ArraySegment<byte> Data { get; private set; }
+        private readonly Communicator _communicator;
+
+        public IncomingRequestFrame(Communicator communicator, ArraySegment<byte> data)
         {
-            // TODO: make this method destructive with memory pooling.
-            return Payload;
+            _communicator = communicator;
+            Data = data;
+            var istr = new InputStream(communicator, data);
+
+            Identity = new Identity(istr);
+
+            // For compatibility with the old FacetPath.
+            string[] facetPath = istr.ReadStringArray();
+            if (facetPath.Length > 1)
+            {
+                throw new InvalidDataException($"invalid facet path length: {facetPath.Length}");
+            }
+            Facet = facetPath.Length == 0 ? "" : facetPath[0];
+            Operation = istr.ReadString();
+            IsIdempotent = istr.ReadOperationMode() != OperationMode.Normal;
+            Context = istr.ReadContext();
+            Payload = Data.Slice(istr.Pos);
+            Encoding = istr.StartEncapsulation();
         }
 
-        public IncomingRequestFrame(InputStream inputStream, Current current)
+        public void ReadEmptyParamList()
         {
-            InputStream = inputStream;
-            Current = current;
+            var istr = new InputStream(_communicator, Payload);
+            istr.StartEncapsulation();
+            istr.EndEncapsulation();
```

Comment:
I think the idea was to not keep this state in the frame, so that the frame can be used concurrently

Reply:
Yes, and a typical use-case here would be several concurrent dispatches on the same `IncomingRequestFrame`.
It's currently allowed but messy and not all that safe, as each retry "rewinds" the InputStream. With a separate InputStream for each attempt (potentially concurrent attempt), there is no issue.

Reply:
Could we consider changing `InputStream` to a struct to minimize the created garbage? This would require to move most of the `InputStream` data members to a separate class which would be instantiated for more advanced decoding: decoding of class, exceptions or endpoints. `InputStream` would mostly only contain a communicator, pos, buffer and a reference to an internal object that would created for the purpose of decoding class/exception/endpoint encapsulations. It's also not clear whether or not we still need to `_mainEncaps` data member, afaict many methods which are using it aren't used anymore... Sounds like it could be a good time to take another pass at simplifying it 😃.

Reply:
I am all for simplifying further InputStream and OutputStream, but this can wait until a new PR.